File: Server/db-client.js
const { MongoClient } = require("mongodb");
const dotenv = require("dotenv");

dotenv.config();
console.log("DB_STRING:", process.env.DB_STRING);

class StableClient {
    constructor() {
        if (StableClient.instance) {
            return StableClient.instance;
        }

        this._client = this.createDbClient();

        StableClient.instance = this;
    }

    createDbClient() {
        const client = new MongoClient(process.env.DB_STRING, {
            useNewUrlParser: true,
            useUnifiedTopology: true,
        });

        client.on("error", async (err) => {
            console.error("Unexpected error on MongoDB client:", err);
            // You can add your own logic for handling errors here
        });

        return client;
    }

    isConnected() {
        return this._client && this._client.isConnected();
    }

    async connect() {
        // If the client is not connected, then connect
        console.log("call to connect");
        if (!this.isConnected()) {
            console.log("not connected, connecting");
            if (!this._client) {
                console.log("no client, creating client");
                this._client = this.createDbClient();
            }
            await this._client.connect();
        }
    }

    async query(collectionName, filter) {
        await this.connect(); // Ensure connection before querying
        console.log("runningQuery");
        const collection = this._client.db().collection(collectionName);
        return collection.find(filter).toArray();
    }

    async end() {
        console.log("call to end connection");
        if (this.isConnected()) {
            console.log("connected, ending connection");
            await this._client.close();
            this._client = null;
        }
    }
}

const client = new StableClient();

module.exports = client;

File: Server/db-client.py
from pymongo import MongoClient
from pymongo.errors import OperationFailure
import os

class StableClient:
    def __init__(self):
        if StableClient.instance:
            return StableClient.instance

        self._client = self.create_db_client()

        StableClient.instance = self

    def create_db_client(self):
        client = MongoClient(os.environ["DB_STRING"], 
                             useNewUrlParser=True, 
                             useUnifiedTopology=True)

        def error_handler(err):
            print(f"Unexpected error on MongoDB client: {err}")
            # You can add your own logic for handling errors here

        client.on_error = error_handler

        return client

    def is_connected(self):
        return self._client and self._client.is_connected

    async def connect(self):
        # If the client is not connected, then connect
        print("call to connect")
        if not self.is_connected():
            print("not connected, connecting")
            if not self._client:
                print("no client, creating client")
                self._client = self.create_db_client()
            await self._client.connect()

    async def query(self, collection_name, filter):
        await self.connect()  # Ensure connection before querying
        print("runningQuery")
        collection = self._client.get_database().get_collection(collection_name)
        return list(collection.find(filter))

    async def end(self):
        print("call to end connection")
        if self.is_connected():
            print("connected, ending connection")
            await self._client.close()
            self._client = None

# Load environment variables from a .env file
from dotenv import load_dotenv
load_dotenv()

client = StableClient()

# Export the instance
client_instance = client._client

# Example usage
async def example_usage():
    await client.connect()
    result = await client.query("your_collection_name", {"your_field": "your_value"})
    print(result)
    await client.end()

if __name__ == "__main__":
    example_usage()
